text/0000-variant-types.md

```diff
@@ -0,0 +1,335 @@
+preserve backwards compatibility, this must remain the case. However, it would
+be convenient for `let x: Foo::Variant1 = Foo::Variant1;` to also be valid.
+
+The type checker must consider multiple types for an enum construction
+expression - both the variant type and the enum type. If there is no further
+information to infer one or the other type, then the type checker uses the enum
+type by default. This is analogous to the system we use for integer fallback or
+default type parameters.
+
+The type of the variants when used as functions must change. Currently they have
+a type which maps from the field types to the enum type:
+
+```rust
+let x: &Fn(i32, &'static str) -> Foo = &Foo::Variant2;
+```
+
+I.e., one could imagine an implicit function definition:
+
+```rust
+impl Foo {
+    fn Variant2(a: i32, b: &'static str) -> Foo { ... }
+}
+```
+
+This would change to accommodate inferring either the enum or variant type,
+imagine
+
+```rust
+impl Foo {
+    fn Variant2<T=Foo>(a: i32, b: &'static str) -> T { ... }
+}
+```
+
+Since we do not allow generic function types, the result type must be chosen
+when the function is referenced:
+
+```rust
+let x: &Fn(i32, &'static str) -> Foo = &Foo::Variant2::<Foo>;
+let x: &Fn(i32, &'static str) -> Foo::Variant2 = &Foo::Variant2::<Foo::Variant2>;
+```
+
+Due to the default type parameter, we remain backwards compatible:
+
+```rust
+let x: &Fn(i32, &'static str) -> Foo = &Foo::Variant2;
+```
+
+Note that this is an innovation. Default type parameters on functions have
+[recently](https://github.com/rust-lang/rust/pull/30724) been feature-gated for
+more consideration. The compiler has never accepted referencing a generic
+function without specifying type parameters, even when there is a default.
+However, I think this should be the expected behaviour. This should be discussed
+further in a separate RFC.
+
+
+## Representation
+
+Enum values have the same representation whether they have enum or variant type.
+That is, a value with variant type will still include the discriminant and
+padding to the size of the largest variant. This is to make sharing
+implementations easier (via coercion), see below.
+
+## Conversions
+
+A variant value may be implicitly coerced to its corresponding enum type (an
+upcast). An enum value may be explicitly cast to the type of any of its variants
+(a downcast). Such a cast includes a dynamic check of the discriminant and will
+panic if the cast is to the wrong variant. Variant values may not be converted
+to other variant types. E.g.,
+
+```
+let a: Foo::Variant1 = Foo::Variant1;
+let b: Foo = a; // Ok
+let _: Foo::Variant2 = a; // Compile-time error
+let _: Foo::Variant2 = b; // Compile-time error
+let _ = a as Foo::Variant2; // Compile-time error
+let _ = b as Foo::Variant2; // Runtime error
+let _ = b as Foo::Variant1; // Ok
+```
+
+## impls
+
+`impl`s may exist for both enum and variant types. There is no explicit sharing
```

Comment:
On Fri, Jan 08, 2016 at 10:06:19AM -0800, Eduard-Mihai Burtescu wrote:
> While @japaric's proposal may be equivalent to the latter option, `let x: [Option<String>; 10] = [None; 10];` would not work with this RFC as written because `None` would infer to `Option<String>` which is not `Copy` - and there is no conversion from `[None<T>; N]` to `[Option<T>; N]` (but there could be? not sure what we can do here).

In the RFC as written, I think something like `[None: None<String>; 32]` would be required?